### ISSUE TITLE
Port twisted.internet.test.test_fdset to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6005]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6005|
|Type|enhancement|
|Created|2012-09-18 16:15:52Z|
|Branch|https://github.com/twisted/twisted/tree/fdset-py3-6005-2|

`twisted.internet.test.test_fdset` is a good starting point for testing the networking parts of the reactor; it should be ported to Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6005 6005
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_fdset_py3_6005_2 branches/fdset-py3-6005-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347984952000000 1347984952000000
changetime__1348760129000000 1348760129000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35741]) Branching to 'fdset-py3-6005'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

(In [35793]) Branching to 'fdset-py3-6005-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

Started build:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/fdset-py3-6005-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  A few minor points:

  1. I don't like the new `timeout=5` in `test_fdset.py`.  Does the test hang instead of failing in a timely manner?  Perhaps that could be fixed.  Is that a pre-existing defect or a result of the Python 3 port?  Maybe we can file a ticket and ignore it for now, since the test passes now.
  1. `RemovingDescriptor` should probably close `read` and `write` in `connectionLost` instead of `doRead`.
  1. Some changes to `_glibbase.py` but some glib-related reactors are skipped when it comes to `test_fdset`.  intentional?

Please merge when the above are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

  1. Filed [#6017](https://github.com/twisted/twisted/issues/6017), removed 5 second timeout.
  1. Added explanation of why it needs to be done in `doRead`, plus bonus test from when I was figuring out why.
  1. I think I did this before I made skipping work in `run-python3-tests`. It's not longer necessary, so I reverted it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [35808]) Merge fdset-py3-6005-2.

Author: itamar
Review: exarkun
Fixes: [#6005](https://github.com/twisted/twisted/issues/6005)

twisted.internet.test.test_fdset now runs on Python 3; this means the reactor now kinda sorta works on Python 3.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set status to `reopened`|
|-|-|

(In [35812]) Revert fdset-py3-6005-2: Failing OS X tests.

Reopens: [#6005](https://github.com/twisted/twisted/issues/6005).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to @exarkun<br>@itamarst set status to `new`|
|-|-|

Started another build, hopefully it'll work now.
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/fdset-py3-6005-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @itamarst|
|-|-|

I guess that seems okay.  Perhaps not using `FileDescriptor` at all in the new test would be better, though, if you feel like it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set status to `closed`|
|-|-|

(In [35821]) Merge fdset-py3-6005-2, this time for reals.

Author: itamar
Review: exarkun
Fixes: [#6005](https://github.com/twisted/twisted/issues/6005)

test_fdset works on Python 3, which probably means the reactor kinda sorta works on Python 3.
